### PR TITLE
feat: add data attr to react-combobox Options

### PR DESCRIPTION
**Pull request by smhigley**
This is in response to a product team request to be able to pass through a custom data value that is available in the `onOptionSelect` callback (something that might be equivalent to the HTML `value` attribute, except our `value` attribute is used to define string values used for the input value and type-to-search functionality).

A `data` prop is added to OptionProps, and is included in the `onOptionSelect` data object.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                             |     Baseline (minified/GZIP) |                           PR |                                                                                                                                                                                                 Change |
| :---------------------------------------------------------------------------------------------------------------------------- | ---------------------------: | ---------------------------: | -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr> | `74.636 kB`<br />`24.186 kB` | `74.688 kB`<br />`24.206 kB` | `52 B` <br />`20 B`  |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr> | `74.236 kB`<br />`24.086 kB` | `74.288 kB`<br />`24.104 kB` | `52 B` <br />`18 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |         Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `187.939 kB`<br />`52.05 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            | `33.359 kB`<br />`11.004 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |   `5.851 kB`<br />`1.964 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/e0a0ab1481d0fb64f2a9804376ad17e4f4f8ea01'>e0a0ab1481d0fb64f2a9804376ad17e4f4f8ea01</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/Avatar-mount_base.html">1329</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/Avatar-mount.html">1285</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/Button-mount_base.html">926</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/Button-mount.html">855</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProvider-mount_base.html">1583</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProvider-mount.html">1463</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-mount_base.html">552</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-mount.html">637</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">586</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">525</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">620</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">630</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/MakeStyles-mount_base.html">1793</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/MakeStyles-mount.html">1717</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/SpinButton-mount_base.html">2447</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24783/perf-test-react-components/SpinButton-mount.html">2414</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by ling1726** [comment]
Is there a need to handle user custom data inside of the component? The component does not actually do anything with this custom data while the user has access to this data from inside their own app/component. 

Users will use their own data to use with the component, so the actual `optionValue` will be known to the user.
They can use this information from the `onOptionSelect` handler to simple get their custom data (that they already have)

```tsx
const MyComponentWithCombobox = ({ itemData }) => {
  const onOptionSelect = (event, data) => {
     // To avoid extra iteration the user could simply create a map as a lookup
     console.log(itemData.find(item => item.itemId === data.optionValue).customData);
  }

  return (
    <Combobox>
      {itemData.map(item => {
        <Option key={item.itemId} />
      })}
    </Combobox>
  )
}
```

This PR will be breaking encapsulation and adding to our API surface, since the actual custom data is **already available to the user** and they **create the releationship to options themselves**. The process of getting the data from the option value is actually quite trivial

File: packages/react-components/react-combobox/src/components/Option/useOption.tsx [hunk 1]
```diff
@@ -31,16 +31,17 @@ function getValueString(value: string | undefined, children: React.ReactNode) {
...
+  const { data, disabled, value } = props;
   const optionRef = React.useRef<HTMLElement>(null);
   const optionValue = getValueString(value, props.children);
 
   // use the id if provided, otherwise use a generated id
   const id = useId('fluent-option', props.id);
 
   // data used for context registration & events
-  const optionData = React.useMemo<OptionValue>(() => ({ id, disabled, value: optionValue }), [
+  const optionData = React.useMemo<OptionValue>(() => ({ id, data, disabled, value: optionValue }), [
```
Comment (ling1726) [inline packages/react-components/react-combobox/src/components/Option/useOption.tsx:42]: you also use `optionData` in an effect, if users don't memoize their `data` it risks running the effect more than you might expect

**Comment by smhigley** [comment]
Closing via #25379

**Pull request closed without merging**